src/mcp3564.c

```diff
@@ -163,21 +163,67 @@ void MCP3561_Reset(SPI_HandleTypeDef *hspi){
 
 /**
  * @brief read 24 Bit left justified ADC register
- * @todo  how to read from other data formats?
+ * @todo  obsolete?
  */
 uint32_t MCP3561_ReadADCData(SPI_HandleTypeDef *hspi){
 	uint8_t val[5] = {0,0,0,0,0};
 	uint8_t cmd[5] = {0,0,0,0,0};
 	cmd[0] = MCP3561_SREAD_DATA_COMMAND;
-	HAL_GPIO_WritePin(SPI1_CS_GPIO_Port, SPI1_CS_Pin, 0);
+	HAL_GPIO_WritePin(MCP3561_CHIP_SELECT_GPIO_Port, MCP3561_CHIP_SELECT_GPIO_Pin, 0);
 	HAL_SPI_TransmitReceive(hspi, cmd, val, 5, 10);
-	HAL_GPIO_WritePin(SPI1_CS_GPIO_Port, SPI1_CS_Pin, 1);
+	HAL_GPIO_WritePin(MCP3561_CHIP_SELECT_GPIO_Port, MCP3561_CHIP_SELECT_GPIO_Pin, 1);
 	uint32_t value = (val[1] << 16) | (val[2] << 8) | val[3];
 	return value;
 }
 
 /*
- * @bug this does not work because it will skip the transaction and write to the chip select pin
+ * @brief read 24 Bit ADC register DATA FORMAT 00 and 01 for MUX Mode.
+ */
+int32_t MCP3561_ReadADCData_24Bit(SPI_HandleTypeDef *hspi){
+	uint8_t val[5] = {0,0,0,0,0};
+	uint8_t cmd[5] = {0,0,0,0,0};
+	cmd[0] = MCP3561_SREAD_DATA_COMMAND;
+	HAL_GPIO_WritePin(MCP3561_CHIP_SELECT_GPIO_Port, MCP3561_CHIP_SELECT_GPIO_Pin, 0);
+	HAL_SPI_TransmitReceive(hspi, cmd, val, 5, 10);
+	HAL_GPIO_WritePin(MCP3561_CHIP_SELECT_GPIO_Port, MCP3561_CHIP_SELECT_GPIO_Pin, 1);
+	int32_t value = (((val[1] & 0x80) >> 7 ) << 31) | ( (val[1] & 0x7F) << 16) | (val[2] << 8) | val[3];

```

Comment:
a lot of magic numbers. i am gonna believe them 🪄

Reply:
Thanks for the contribution. A lot of changes but seems to be fine. Unfortunately, I don't have the hardware anymore, so I cannot test it physically.